tensorflow/python/ops/image_ops_impl.py

```diff
@@ -182,8 +182,81 @@ def _CheckAtLeast3DImage(image, require_static=True):
     return []
 
 
-def fix_image_flip_shape(image, result):
-  """Set the shape to 3 dimensional if we don't know anything else.
+def _EnsureTensorIs4D(image):
+  """Converts `image` to a 4-D Tensor if it is not already one.
+
+  Args:
+    image: 4-D Tensor of shape `[batch, height, width, channels]` or
+           3-D Tensor of shape `[height, width, channels]`.
+  Raises:
+    ValueError: if image is not a 3-D or 4-D Tensor.
+
+  Returns:
+    If `image` was 4-D, a 4-D float Tensor of shape
+    `[batch, width, height, channels]`
+    If `image` was 3-D, a 4-D float Tensor of shape
+    `[1, width, height, channels]`
+  """
+  original_shape = image.get_shape()
+  is_batch = True
+  if original_shape.ndims == 3:
+    is_batch = False
+    image = array_ops.expand_dims(image, 0)
+  elif original_shape.ndims is None:
+    is_batch = False
+    image = array_ops.expand_dims(image, 0)
+    image.set_shape([None] * 4)
+  elif original_shape.ndims != 4:
+    raise ValueError('\'image\' must have either 3 or 4 dimensions.')
+
+  return (image, is_batch)
+
+def _flip_image(image, axis, random=False, seed=None):
+  """
+  Flips image(s) around a given axis.
+
+  Args:
+    image:  4-D Tensor of shape `[batch, height, width, channels]` or
+            3-D Tensor of shape `[height, width, channels]`.
+    axis:   A Python integer representing the axis on which the image(s)
```

Comment:
As mentioned [here](https://github.com/tensorflow/tensorflow/pull/14854#discussion_r155128239):

One strange thing about this refactor is that it forces the caller to provide an axis relative to a 4-D Tensor, even if they've passed in a 3-D image.

For example `random_flip_up_down()` needs to provide an axis, but doesn't know whether or not the image it's working with is 3-D or a 4-D batch of images. Currently I just pass in `1` with the knowledge that a single image will be extended to 4 dimensions.

Reply:
That's a good point -- but I think that's ok. It's made clear in the argument description, and in the end, this is a private utility function, so I'm not too worried about it.

Reply:
I guess if you wanted to, you could make a symbolic argument: direction='UP_DOWN' or 'LEFT_RIGHT' and interpret that. But again, the function is private, I don't think it's worth it.

Reply:
👍